File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        cnt = Counter(nums)
        buckets = [[] for _ in range(len(nums) + 1)]

        for key in cnt.keys():
            key_count = cnt[key]
            buckets[key_count].append(key)
        
        output = []
        for i in range(len(buckets) - 1, -1, -1):
            for num in buckets[i]:
                if k > 0:
                    output.append(num)
                    k -= 1
        
        return output

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> cnt;
        vector<vector<int>> buckets(nums.size() + 1);
        
        for (int num : nums) {
            cnt[num]++;
        }

        for (const auto& pair : cnt) {
            buckets[pair.second].push_back(pair.first);
        } 

        vector<int> result;
        for (int i = buckets.size() - 1; i > 0 && k > 0; i--) {
            for (int num: buckets[i]) {
                if  (k > 0) {
                    result.push_back(num);
                    k--;
                }
            }
        }

        return result;
    }
};